### PR TITLE
TGM3 grading (complete)

### PR DESCRIPTION
**Pull request by CreepercraftYT**

### CODE REVIEW

**Comment by CreepercraftYT** [comment]
this should work, btw, since COOLs skip a section (Speed-wise), can you add them (and REGRETs too)?

**Comment by Villadelfia** [comment]
> this should work, btw, since COOLs skip a section (Speed-wise), can you add them (and REGRETs too)?

I'm not sure how I'd implement that. The way levels are set up currently doesn't really support that, have a look in `level.asm` to see how speed curves are done currently.

I'll get to the review soon, sorry. Slept a literal 24 hours.

**Comment by CreepercraftYT** [comment]
added multipliers for TGM3 and a grading system for Chill mode

File: src/include/globals.asm [hunk 1]
```diff
@@ -393,6 +393,12 @@ DEF NLEVEL_THOUSANDS        EQU 0
...
-
-
+DEF LEVEL_MULT_1A           EQU $02
+DEF LEVEL_MULT_1B           EQU $50
+DEF LEVEL_MULT_2A           EQU $05
+DEF LEVEL_MULT_3A           EQU $07
+DEF LEVEL_MULT_3B           EQU $50
+
+; Bone Skin
+DEF BONE                EQU $68
```
Comment (Villadelfia) [inline src/include/globals.asm]: If it's a tile, it should be titled "TILE_XXX"

File: src/grading.asm [hunk 2]
```diff
@@ -213,24 +222,29 @@ SECTION "Grading Functions", ROM0
...
-    ; TGM1 and DMGT are the exceptions.
+    ; TGM1, TGM3, CHIL and DMGT are the exceptions.
     ld a, [wSpeedCurveState]
     cp a, SCURVE_TGM1
     jr z, .grade9start
     cp a, SCURVE_DMGT
     jr z, .grade9start
+    cp a, SCURVE_TGM3
+    jr z, .grade9start
+    cp a, SCURVE_CHIL
```
Comment (Villadelfia) [inline src/grading.asm:342]: Not sure if I'm personally a fan of having grades in CHIL mode at all. It's supposed to be a chill mode after all.

File: src/grading.asm [hunk 3]
```diff
@@ -861,4 +875,329 @@ UpdateGradeSHIR:
...
+    cp a, LEVEL_MULT_3B
+    ; If C is less than 50, we didn't reach level 750 yet, but we are for sure in the 7** Section, load the corresponding offset.
+    jp .under750
+    ; If C is equal or greater than 50, then congrats!, load the corresponding offset
+    ld b, 0
+    ld c, 2 
+    add hl, bc
+    ld a, [hl]
+    jp .Multiply
+.under750
```
Comment (Villadelfia) [inline src/grading.asm:1255]: You seem to choose .UpperCase and .lowercase almost randomly. Best to choose one and stick with it.

File: src/grading.asm [hunk 4]
```diff
@@ -861,4 +875,329 @@ UpdateGradeSHIR:
...
+    ld c, a
+    add hl, bc
+    ld a, [hl] ; Load the rate into a...
+    ld [wDecayRate], a ; ... and then into the timer
+    ret
+.lpoints
+    ld [wInternalGradePoints], a
+    ret 
+
+UpdateGradeCHIL:
```
Comment (Villadelfia) [inline src/grading.asm:1584]: See earlier comment about CHIL grading.

File: src/field.asm [hunk 5]
```diff
@@ -1764,7 +1764,16 @@ FieldProcess::
 
 .postghost
     ; If the lock delay is at the highest value, draw the piece normally.
-    ldh a, [hCurrentPiece]
+    ld a, [wSpeedCurveState]
+    cp a, 4
```
Comment (Villadelfia) [inline src/field.asm]: Magic number

File: src/field.asm [hunk 6]
```diff
@@ -3976,7 +3985,16 @@ BigFieldProcess::
...
```
Comment (Villadelfia) [inline src/field.asm]: Magic number

**Comment by CreepercraftYT** [comment]
this is not the newest code btw, i'll fix the magic numbers

File: src/grading.asm (hunk 2 above)
Comment (CreepercraftYT) [reply src/grading.asm:342]: well, it's chill, an easy-to-get GM

**Comment by CreepercraftYT** [comment]
i haven't finished the regrets, so for now, either merge but don't build or just don't merge

**Comment by CreepercraftYT** [comment]
note: a section timer is required

File: src/level.asm [hunk 7]
```diff
@@ -213,11 +213,6 @@ SpecialLevelInit:
     ; Increment level and speed up if necessary. Level increment in E.
     ; Levels may only increment by single digits.
 LevelUp::
-    ; Return if our level is hard locked.
-    ld a, [wLockLevel]
-    cp a, $FF
-    ret z
```
Comment (Villadelfia) [inline src/level.asm:227]: This is required to keep level locked to 999, 300, 1300, or whatever the max level is of a mode. Why was this removed?

File: src/include/globals.asm [hunk 8]
```diff
@@ -74,7 +74,6 @@ CHARMAP "c", 209
 CHARMAP "d", 210
 CHARMAP "m", 211
 CHARMAP ":", 212
-CHARMAP ".", 254
```
Comment (Villadelfia) [inline src/include/globals.asm]: Why was this removed?

File: README.md [hunk 9]
```diff
@@ -61,6 +60,15 @@ Select between several speed curves including the DMGTRIS default speed curve, T
 
 Note that all modes use the same scoring.
 
+Some curves have their own grading system
+
+- DMGT: It's like the TAP Grading system, but you keep the remainder of the grade points and you need to reach 150 points instead of 100
```
Comment (Villadelfia) [inline README.md]: I'll write these in my own words.

File: src/medals.asm
Comment (Villadelfia) [inline src/medals.asm:1]: Probably best to leave out stubbed files. Also, you forgot an ENDC at the end of the file.

File: src/res/other_data.inc [hunk 10]
```diff
@@ -286,7 +286,7 @@ sTGM1SpeedCurve::
     db 1, 192
     db 30, 30, 16, 30, 41
 
-    dw $0247, 247, $0300
```
Comment (Villadelfia) [inline src/res/other_data.inc:289]: Make sure to keep the version here with the data typo fixed!

File: src/res/other_data.inc [hunk 11]
```diff
@@ -538,7 +537,7 @@ sTGM3SpeedCurve::
     db 1, 192
     db 27, 27, 16, 30, 40
 
-    dw $0247, 247, $0300
```
Comment (Villadelfia) [inline src/res/other_data.inc:541]: Make sure to keep the version here with the data typo fixed!

File: src/res/other_data.inc [hunk 12]
```diff
@@ -594,7 +593,7 @@ sTGM3SpeedCurve::
     db 20, $00
     db 8, 8, 8, 17, 6
 
-    dw $1100, 1100, $1200
```
Comment (Villadelfia) [inline src/res/other_data.inc:597]: Make sure to keep the version here with the data typo fixed!

File: src/res/other_data.inc [hunk 13]
```diff
@@ -619,19 +618,19 @@ sDEATSpeedCurve::
     db 20, $00
     db 18, 14, 12, 30, 12
 
-    dw $0100, 100, $0200
```
Comment (Villadelfia) [inline src/res/other_data.inc:622]: Make sure to keep the version here with the data typo fixed!

File: src/res/other_data.inc [hunk 14]
```diff
@@ -619,19 +618,19 @@ sDEATSpeedCurve::
...
+    dw $0100, 0, $0200
     db 20, $00
     db 14, 8, 12, 26, 6
 
-    dw $0200, 200, $0300
```
Comment (Villadelfia) [inline src/res/other_data.inc:626]: Make sure to keep the version here with the data typo fixed!

File: src/res/other_data.inc [hunk 15]
```diff
@@ -619,19 +618,19 @@ sDEATSpeedCurve::
...
+    dw $0200, 0, $0300
     db 20, $00
     db 14, 8, 11, 22, 6
 
-    dw $0300, 300, $0400
```
Comment (Villadelfia) [inline src/res/other_data.inc:630]: Make sure to keep the version here with the data typo fixed!

File: src/res/other_data.inc [hunk 16]
```diff
@@ -619,19 +618,19 @@ sDEATSpeedCurve::
...
+    dw $0300, 0, $0400
     db 20, $00
     db 8, 8, 10, 18, 6
 
-    dw $0400, 400, $0500
```
Comment (Villadelfia) [inline src/res/other_data.inc:634]: Make sure to keep the version here with the data typo fixed!

File: src/res/other_data.inc [hunk 17]
```diff
@@ -803,14 +802,6 @@ sMYCOSpeedCurve::
     db 20, $00
     db 20, 20, 8, 30, 20
 
-    dw $0200, 200, $0300
```
Comment (Villadelfia) [inline src/res/other_data.inc:806]: Make sure to keep the version here with the data typo fixed!

File: src/res/other_data.inc [hunk 18]
```diff
@@ -2175,7 +2166,7 @@ sGameplayTilesC::
     DB $18,$18,$2C,$34,$62,$5E,$76,$6E
     DB $24,$3C,$24,$3C,$18,$18,$00,$00
     DB $00,$00,$00,$00,$00,$00,$00,$00
-    DB $00,$00,$30,$30,$30,$30,$00,$00
```
Comment (Villadelfia) [inline src/res/other_data.inc:2178]: Merge error?

File: src/res/other_data.inc [hunk 19]
```diff
@@ -1756,7 +1747,7 @@ sGameplayTilesM::
...
```
Comment (Villadelfia) [inline src/res/other_data.inc:1759]: Merge error?

File: src/res/other_data.inc [hunk 20]
```diff
@@ -911,8 +902,8 @@ sSharedTiles::
     DB $03,$03,$03,$03,$0F,$0F,$00,$00
     DB $00,$00,$00,$00,$00,$00,$3C,$3C
     DB $00,$00,$00,$00,$00,$00,$00,$00
-    DB $18,$18,$18,$18,$08,$08,$10,$10
```
Comment (Villadelfia) [inline src/res/other_data.inc:914]: Merge error?

File: src/res/other_data.inc [hunk 21]
```diff
@@ -701,7 +700,7 @@ sSHIRSpeedCurveEnd::
     dw $FFFF
 
 sSHIRSpeedCurveSpecialData::
-    dw 900     ; When do we get bones?
+    dw 1000    ; When do we get bones?
```
Comment (Villadelfia) [inline src/res/other_data.inc:703]: Good catch.

File: src/sprites.asm
Comment (Villadelfia) [inline src/sprites.asm:1]: The bones in sprites were generalized, this file can be left unmodified.

File: src/time.asm [hunk 22]
```diff
@@ -121,6 +131,35 @@ CheckTorikan::
     ld a, $FF
     ret
 
+CheckCOOL_REGRET::
```
Comment (Villadelfia) [inline src/time.asm:134]: Looks good.

File: src/time.asm [hunk 23]
```diff
@@ -137,7 +176,12 @@ HandleTimers::
     ret z
     cp a, MODE_PRE_GAME_OVER
     ret z
-
+    ld a, [hCLevel+CLEVEL_ONES]
```
Comment (Villadelfia) [inline src/time.asm:179]: This will reset the section time every frame where the ones digit <6. This means XX00-XX05, but also XX11-15, XX21-XX25 etc. You probably need to check for the tens digit being 0 and, if the thousands digit isn't zero, that the hundreds digit is zero.

File: src/time.asm [hunk 24]
```diff
@@ -195,7 +239,56 @@ HandleTimers::
     ld a, [wMinutes]
     inc a
     ld [wMinutes], a
-    ret
 
+HandleSectionTimers::
+    ldh a, [hFrameCtr]
```
Comment (Villadelfia) [inline src/time.asm]: This also happens in HandleTimers, now the frame counter will be incremented twice per frame and every frame will be considered odd. Remove this block please from L244-L248.

File: src/grading.asm (hunk 2 above)
Comment (Villadelfia) [reply src/grading.asm:342]: I'll keep CHIL and MYCO gradeless in the original version, but you are free to give those modes grades in your fork.

File: src/grading.asm [hunk 25]
```diff
@@ -365,45 +463,7 @@ DrawGradeProgressDMGT::
     ret
 
 UpdateGradeDMGT::
-    ; Check if the torikan hasn't been calculated.
```
Comment (Villadelfia) [inline src/grading.asm]: Did you accidentally remove the DMGT grading?

File: src/grading.asm [hunk 26]
```diff
@@ -1059,38 +1108,539 @@ UpdateGradeSHIR:
...
+    ld c, a ; We will add from C
+:   add a, c
+    dec b
+    jr nz, :-
+    ; Finally!, we can now load the points, right?, NO!, there is yet another multiplier...
+    ; We have to keep the points safe again...
+    ld e, a
+.levelmultiplier
+    ; Make HL point to the table that contains the level multipliers
+    ld hl, sTGM3LevelMultiplier
```
Comment (Villadelfia) [inline src/grading.asm:1224]: I don't think you ever use this table?

File: src/grading.asm [hunk 27]
```diff
@@ -1059,38 +1108,539 @@ UpdateGradeSHIR:
...
+    ; If not, continue
+    ; Do we have 100 Grade Points?
+    ld a, [wInternalGradePoints]
+    cp a, 100
+    ret c ; If the Internal Grade Points is less than 100, return, we don't have to change our Grade
+    xor a, a ; Reset the points to 0 and increase the internal grade
+    ld [wInternalGradePoints], a
+    ld a, [wInternalGrade]
+    inc a
+    ld [wInternalGrade], a
```
Comment (Villadelfia) [inline src/grading.asm:1313]: It is a good idea to leave a comment when one function falls through to the next, so that I know it's not a mistake.

File: src/res/other_data.inc [hunk 28]
```diff
@@ -607,10 +606,10 @@ sTGM3SpeedCurveEnd::
 
 sTGM3SpeedCurveSpecialData::
     dw $FFFF    ; When do we get bones?
-    dw 1299     ; When do blocks turn invisible?
```
Comment (Villadelfia) [inline src/res/other_data.inc:610]: I haven't seen any code that skips sections, is it okay to shorten the curve?

**Review by Villadelfia — CHANGES_REQUESTED** [review]

File: src/time.asm (hunk 23 above)
Comment (CreepercraftYT) [reply src/time.asm:179]: oops

File: src/level.asm (hunk 7 above)
Comment (CreepercraftYT) [reply src/level.asm:227]: ummm... hold on

File: src/grading.asm (hunk 26 above)
Comment (CreepercraftYT) [reply src/grading.asm:1224]: the level multiplier one is used

File: src/grading.asm (hunk 25 above)
Comment (CreepercraftYT) [reply src/grading.asm]: i didn't delete it, i was also confused when i uploaded the files

File: src/res/other_data.inc (hunk 18 above)
Comment (CreepercraftYT) [reply src/res/other_data.inc:2178]: i didn't touch that part of the file

File: src/res/other_data.inc (hunk 19 above)
Comment (CreepercraftYT) [reply src/res/other_data.inc:1759]: i didn't touch that part of the file

File: src/res/other_data.inc (hunk 20 above)
Comment (CreepercraftYT) [reply src/res/other_data.inc:914]: i didn't touch that part of the file

File: src/time.asm (hunk 24 above)
Comment (CreepercraftYT) [reply src/time.asm]: ok

**Comment by CreepercraftYT** [comment]
a lot of messed up stuff, i'll try to fix everything

File: src/res/other_data.inc (hunk 28 above)
Comment (CreepercraftYT) [reply src/res/other_data.inc:610]: until we find a way to skip sections, yes

File: src/include/globals.asm (hunk 8 above)
Comment (CreepercraftYT) [reply src/include/globals.asm]: accidental

File: README.md (hunk 9 above)
Comment (CreepercraftYT) [reply README.md]: ok

File: src/medals.asm
Comment (CreepercraftYT) [reply src/medals.asm:1]: ok

File: src/time.asm (hunk 23 above)
Comment (CreepercraftYT) [reply src/time.asm:179]: btw, that's not the thing that resets the timer

**Comment by CreepercraftYT** [comment]
i think i didn't miss anything, can you do the data related stuff for me?, i don't know what the values should be

**Comment by Villadelfia** [comment]
Merged

**Comment by CreepercraftYT** [comment]
:D
